conans/client/toolchain/cmake.py

```diff
@@ -0,0 +1,184 @@
+            set(CMAKE_EXE_LINKER_FLAGS_INIT "${CONAN_EXE_LINKER_FLAGS}" CACHE STRING "" FORCE)
+        else()
+            # message(">>>> TRY COMPILE")
+        endif()
+        
+        conan_set_compiler()
+
+
+        {#
+        # Host machine        
+        set(CMAKE_SYSTEM_NAME {{host.os}})
+        set(CMAKE_SYSTEM_VERSION {{host.os_version}})
+        set(CMAKE_SYSTEM_PROCESSOR {{host.arch}})
+
+        # Build machine (only if different)
+        set(CMAKE_HOST_SYSTEM_NAME {{build.os}})
+        set(CMAKE_HOST_SYSTEM_VERSION {{build.os_version}})
+        set(CMAKE_HOST_SYSTEM_PROCESSOR {{build.arch}})
+        
+        set(CMAKE_C_COMPILER {{c_compiler}})
+        set(CMAKE_CXX_COMPILER {{cxx_compiler}})
+        #}
+    """)
+
+    def __init__(self, conanfile,
+                 generator=None,
+                 cmake_system_name=True,
+                 parallel=True,
+                 build_type=None,
+                 toolset=None,
+                 make_program=None,
+                 msbuild_verbosity="minimal",
+                 # cmake_program=None,  # TODO: cmake program should be considered in the environment
+                 generator_platform=None
+                 ):
+        self._conanfile = conanfile
+        del conanfile
+
+        generator = generator or get_generator(self._conanfile.settings)
+        self._context = {
+            "conan_adjustements_cmake": self.conan_adjustements.replace("\\", "/"),
+            "conan_project_include_cmake": self.conan_project_include.replace("\\", "/"),
+            #"build_type": build_type or self._conanfile.settings.get_safe("build_type"),
+            #"generator": generator,
+            "generator_platform": generator_platform or
+                                  get_generator_platform(self._conanfile.settings, generator),
+            # "parallel": parallel,  # TODO: This is for the --build
+            # '-Wno-dev'  # TODO: Can I add this to a CMake variable?
+            "toolset": toolset or get_toolset(self._conanfile.settings)
+        }
+
+        builder = CMakeDefinitionsBuilder(self._conanfile,
+                                          cmake_system_name=cmake_system_name,
+                                          make_program=make_program, parallel=parallel,
+                                          generator=generator,
+                                          set_cmake_flags=False,
+                                          forced_build_type=build_type,
+                                          output=self._conanfile.output)
+        self.definitions = builder.get_definitions()
+        build_type = self.definitions.pop('CMAKE_BUILD_TYPE', None)
+        if build_type:
+            self._context.update({
+                "CMAKE_BUILD_TYPE": build_type
+            })
+
+        # Some variables can go to the environment
+        # TODO: Do we need this or can we move it to environment stuff
```

Comment:
Quick and out loud question related to your TODO comment: Might the generators be able to declare environment also? the pkg_config generator generates a bunch of `pc` files in the install_dir, but it is convenient for the consumers to have that install dir in the path, that's why all of this.